### PR TITLE
4254 try to process all accepted requests before unblock Server.Shutdown()

### PR DESCRIPTION
**Pull request by akaspin**
`graceful` uses builtin Go `http.Server` to accept requests. It provides "stop" channels. Here we have naming issue (in `graceful`) because these channels means that server stops accepting requests *but not* means that all already accepted requests are completed. 

This PR adds middleware to maintain `sync.WaitGroup`. `Server.Shutdown()` blocks till all requests processed or `ShutdownTimeout` is reached.

### CODE REVIEW

File: gin.go [hunk 1]
```diff
@@ -54,3 +55,12 @@ func ginRecovery(name string) gin.HandlerFunc {
 		c.Next()
 	}
 }
+
+func ginRequestsBlocker(wg *sync.WaitGroup) gin.HandlerFunc {
+	return func(_ *gin.Context) {
+		wg.Add(1)
+		defer func() {
+			wg.Done()
+		}()
```
Comment (mbertschler) [inline gin.go]: Saw this in the chat. What is the purpose of the defer?
https://github.com/gin-gonic/gin#custom-middleware
Maybe this needs `ctx.Next()` like
```go
wg.Add(1)
ctx.Next()
wg.Done()
```

Comment (akaspin) [reply gin.go]: No `Next` was intended during investigation surgery :-) 
`defer` guarantees that `Done()` will be called.

File: server.go [hunk 2]
```diff
@@ -101,6 +104,7 @@ func (s *Server) Init() error {
 	gin.SetMode("release")
 	s.Engine = gin.New()
 	s.Engine.Use(
+		ginRequestsWaiter(&(s.requestsWg)),
```
Comment (mbertschler) [inline server.go]: extra () not needed

File: server.go [hunk 3]
```diff
@@ -122,6 +126,18 @@ func (s *Server) Shutdown(os.Signal) {
...
+	allRequestsServedChan := make(chan struct{})
+	go func() {
+		s.requestsWg.Wait()
+		close(allRequestsServedChan)
+	}()
+	select {
+	case <-allRequestsServedChan:
+		s.log.Info("all requests processed")
+	case <-time.After(s.ShutdownTimeout):
+		_ = s.log.Error(fmt.Errorf("shutdown timeout reached"), "remained unprocessed requests")
```
Comment (mbertschler) [inline server.go]: `_ =` is not really necessary

Comment (akaspin) [reply server.go]: lint. `cue.Error()` returns error. error handling is important (60% of our bugs originates from ruby-style no-error code). even on this level.

File: server.go (hunk 2 above)
Comment (akaspin) [reply server.go]: old school :-D

**Pull request merged**